bugfix: include configuration ignoring files with the same basename

RikkiButler20 opened the pull request:
Fixes [#17992](https://github.com/spack/spack/issues/17992)

When using the include: field inside a spack.yaml for an environment, if two of the included configurations have the same basename, only the first file with the same basename is "included". I would have expected both to be included

RikkiButler20 committed:
Recreating bug

RikkiButler20 committed:
Merge branch 'develop' of github.com:spack/spack into bugfix/include-config

RikkiButler20 committed:
Merge branch 'develop' of github.com:spack/spack into bugfix/include-config

RikkiButler20 committed:
Merge branch 'develop' of github.com:spack/spack into bugfix/include-config

RikkiButler20 committed:
Bad solution
Adding the time to the configuration name

RikkiButler20 committed:
Fix flake8 mistakes

RikkiButler20 committed:
Use the config path instead of the basename

RikkiButler20 committed:
Flake8 warnings

RikkiButler20 committed:
Merge branch 'develop' of github.com:spack/spack into bugfix/include-config

RikkiButler20 committed:
Removing changes to recreate bug

becker33 commented on lib/spack/spack/environment.py:
```diff
@@ -830,7 +830,7 @@ def included_config_scopes(self):
             elif os.path.exists(config_path):
                 # files are assumed to be SingleFileScopes
                 base, ext = os.path.splitext(os.path.basename(config_path))
```
```suggestion
```
If you're not using `base` or `ext` then I think this line can go.

becker33 reviewed (commented):
Is `config_path` the relative or absolute file path?

RikkiButler20 committed:
Update lib/spack/spack/environment.py

Removing unused variables

Co-authored-by: Greg Becker <becker33@llnl.gov>

RikkiButler20 commented:
`config_path` is an absolute file path

RikkiButler20 committed:
Test
Making sure if therer are 2 include config files with the same basename they are both implemented

RikkiButler20 committed:
Flake8 indentation fix

becker33 commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -547,6 +547,55 @@ def test_with_config_bad_include(env_deactivate, capfd):
+    with e:
+        e.concretize()
+
+    mpileaks_found = False
+    libelf_found = False
+
+    for x in e._get_environment_specs():
+        if mpileaks_found and libelf_found:
+            break
+        if x.satisfies('mpileaks@2.2'):
+            mpileaks_found = True
+        elif x.satisfies('libelf@0.8.10'):
+            libelf_found = True
+    assert(mpileaks_found and libelf_found)
```
I think the assertion logic at the end of the test could be cleaned up. Something like
```suggestion
    for expected in ('mpileaks@2.2', 'libelf@0.8.10'):
        assert any(cspec.satisfies(expected) for _, cspec in e.concretized_specs())
```

Another option would be to add the `recurse_dependencies=False` option to `_get_environment_specs`, that way there wouldn't be any additional specs to iterate over.

becker33 replied:
The test as written could also technically pass for the wrong reason. `mpileaks` depends on `libelf`, so if the `mpileaks` dependency on `libelf` used version 0.8.10 but the top-level spec did not, the test should fail. I think in this case it would fail as-is, but only because `libelf` is before `mpileaks` in the spec list, which is more fragile than it should be.

RikkiButler20 replied:
I'm going with the `recurse_dependencies=False`.

becker33 reviewed (changes requested):
The bugfix looks good. One request on the test.

RikkiButler20 committed:
Edit test assert
